rpc-broker/src/msg.c

```diff
@@ -0,0 +1,286 @@
+/*
+ Copyright (c) 2018 AIS, Inc.
+
+ This program is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2 of the License, or
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#include "rpc-broker.h"
+
+
+/*
+ * All requests are handled by this function whether they are raw requests
+ * or Websocket requests.  For every policy rule listed either in the
+ * /etc/rpc-broker.policy file or the domain-specific rules listed in the
+ * xenclient database, each is passed to `filter` to determine whether or not
+ * the message is dropped or passed through.
+ */
+int broker(struct dbus_message *dmsg, int domid)
+{
+    int policy, current_rule_policy, domains;
+    struct etc_policy etc;
+
+    int i, j;
+    char req_msg[1024] = { '\0' };
+    char *uuid;
+
+    if (!dmsg) {
+        DBUS_BROKER_WARNING("Invalid args to broker-request %s", "");
+        return -1;
+    }
+
+    if (!dbus_broker_policy) {
+        DBUS_BROKER_WARNING("No policy in place %s", "");
+        return -1;
+    }
+
+    // deny by default
+    policy = 0;
+    // sets the current-rule-policy to default
+    current_rule_policy = policy;
+
+    etc = dbus_broker_policy->etc;
+
+    for (i=0; i < etc.count; i++) {
+        current_rule_policy = filter(&(etc.rules[i]), dmsg, domid);
+        /*
+         *  1 = the rule matched the request and rule's policy is allow
+         *  0 = the rule matched the request and rule's policy is deny
+         * -1 = the rule did *not* match the request
+         *
+         *  The filter is set up as last match, with the default being "deny".
+         *  Iterating through all the rules, if a rule matches a request, then
+         *  update the "policy" to fit with the "current-rule-policy"
+         *  (whether thats allow or deny).  If there is no match (-1) then the
+         *  "policy" remains unchanged, remaining set as it was before the
+         *  rule being checked.
+         */
+        policy = current_rule_policy != -1 ? current_rule_policy : policy;
+    }
+
+    domains = dbus_broker_policy->domain_count;
+    if (domain_uuids[domid])
+        uuid = domain_uuids[domid];
+    else {
+        uuid = get_uuid_from_domid(domid);
+        domain_uuids[domid] = uuid;
+    }
+
+    for (i=0; i < domains; i++) {
+        if (uuid && !strcmp(dbus_broker_policy->domains[i].uuid_db_fmt, uuid)) {
```

Comment:
Looping through all the domains until you find the one with the right uuid and then do stuff and break...
Maybe a `get_domain_policy(char *uuid)` would simplify this.